libbpf-tools/mysqld_qslower.c

```diff
@@ -0,0 +1,214 @@
+const char argp_program_doc[] =
+"Show MySQL server queries slower than a threshold.\n"
+"\n"
+"USAGE: mysqld_qslower [-h] [-p PID] [-l LAT]\n"
+"\n"
+"EXAMPLES:\n"
+"    mysqld_qslower -p 1215             # only trace PID 1215\n"
+"    mysqld_qslower -p 1215 -l 10       # trace query slower than 10ms\n";
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Process ID to trace" },
+	{ "lat", 'l', "LAT", 0, "Min latency to trace, in ms (default 1)" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	long n;
+
+	switch (key) {
+	case 'p':
+		errno = 0;
+		n = strtol(arg, NULL, 10);
+		if (errno || n <= 0) {
+			fprintf(stderr, "Invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		target_pid = n;
+		break;
+	case 'l':
+		errno = 0;
+		n = strtol(arg, NULL, 10);
+		if (errno || n < 0) {
+			fprintf(stderr, "Invalid LAT: %s\n", arg);
+			argp_usage(state);
+		}
+		lat_ms = n;
+		break;
+	case 'v':
+		verbose = true;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct event *e = data;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+
+	printf("%-8s %-7d %-8.3f %s\n", ts, e->pid, (double)e->lat_ns / 1000 / 1000, e->query);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	fprintf(stderr, "lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+static int attach_usdt(struct mysqld_qslower_bpf *obj)
+{
+	char path[PATH_MAX];
+
+	snprintf(path, sizeof(path), "/proc/%d/exe", target_pid);
+
+	obj->links.query_start = bpf_program__attach_usdt(obj->progs.query_start, target_pid, path,
+							  "mysql", "query__start", NULL);
+	if (!obj->links.query_start)
+		return -errno;
+
+	obj->links.query_done = bpf_program__attach_usdt(obj->progs.query_done, target_pid, path,
+							 "mysql", "query__done", NULL);
+	if (!obj->links.query_done)
+		return -errno;
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	LIBBPF_OPTS(bpf_object_open_opts, open_opts);
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct mysqld_qslower_bpf *obj;
+	struct perf_buffer *pb = NULL;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_strict_mode(LIBBPF_STRICT_ALL);
+	libbpf_set_print(libbpf_print_fn);
+
+	err = ensure_core_btf(&open_opts);
```

Comment:
you can write this tool without CO-RE, right? Latest libbpf doesn't require CO-RE for USDT support either. So let's have an example of tracing tool which doesn't depend on vmlinux BTF?